### PR TITLE
Expose Sink as constructor type.

### LINKED ISSUE #38: Confirm we don't buffer more than we read
Might need to change our APIs to let the user provide a BufferedSource, since right now we might consume more data than we should.

https://github.com/square/moshi/pull/33

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #38.

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java [hunk 1]
```diff
@@ -180,7 +181,11 @@
   /**
    * Creates a new instance that writes a JSON-encoded stream to {@code sink}.
    */
-  public JsonWriter(BufferedSink sink) {
+  public JsonWriter(Sink sink) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: Do we offer the necessary APIs to flush or emit this?

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 2]
```diff
@@ -26,12 +28,25 @@
...
   public final T fromJson(String string) throws IOException {
     return fromJson(new JsonReader(string));
   }
 
   public abstract void toJson(JsonWriter writer, T value) throws IOException;
 
+  public final void toJson(Sink sink, T value) throws IOException {
+    JsonWriter writer = new JsonWriter(sink);
+    toJson(writer, value);
+    writer.close();
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: Closing is an odd choice. What if you're writing JSON to a file that contains mixed data?

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java (hunk 1 above)
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: Yes. `JsonWriter` is a `Flushable` which flushes its `BufferedSink` wrapper.

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java (hunk 2 above)
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: Switched to `.flush()`.

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java (hunk 1 above)
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java]: I can add Javadoc to that effect if you think it's likely to be a point of user confusion / problems.

**Comment by JakeWharton** [comment]
Updated to address #38 with test.

**Comment by swankjesse** [comment]
Dig.

**Pull request merged**